Try: decouple test and live mode customers

luizreis opened the pull request:
Fixes #

This PR is an attempt at validating and fixing the concerns raised in https://github.com/Automattic/woocommerce-payments/pull/899#discussion_r490269452. 

New customers and payment methods are created in the correct metadata according to the current test mode setting. The payment methods displayed in the checkout are filtered to match the customer and test mode setting, preventing customers from using unavailable payment methods.

For older customers, this PR attempts to add the customer present in the old `wcpay_customer_id` metadata to `wcpay_customer_id_test` or `wcpay_customer_id_live` based on the current test mode setting. While that works for most cases, it is possible that a customer was created in another mode and shoppers would get an error.

To prevent and fix that, we catch the `resource_missing` exception and move the customer to the right mode if the error message indicates that the customer exists in the other mode.

#### Changes proposed in this Pull Request

* Decouple test and live customers into two separate meta keys
* Filter a user's tokens based on the `customer_id` and `test_mode`
* Migrate old customers to the new meta keys
* Check for customers saved in the wrong environment when fetching their payment methods
* Check for customers saved in the wrong environment when adding new cards from My account

<!--
- [x] Store test and live mode customer IDs in separate options
- [x] Add customer ID to token when creating it
- [x] Remove tokens from other customer IDs before displaying them
- [x] Add test mode information to tokens
- [x] ~Tweak customer creation logic to ensure a user always has a customer~
    * ~Perhaps create a customer in `get_customer_id_by_user_id` if there's no customer in the meta~
    * ~Continue to update the customer during checkout~
    * Will check to see how we do it on stripe: Stripe does it similarly to what we do now, so I don't think we should change it in this PR
- [x] Figure out a migration path for existing customers and tokens
    * ~We'd need to check in which mode (test or live) they were created, which could not match the current mode~
    * Or we could assume that they're live. It shouldn't be too bad to lose some test customers
        * By going that route we still need to figure out a way to migrate the payment methods created before this PR
- [ ] ~Take into account the mode in which a subscription was created when processing off-site payments~
    * ~If a subscription was set up in live mode, but the site is in test mode when the renewal happens, it could lead to issues charging the saved token~
    * This should be handled in another issue and PR
-->

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

There are a few scenarios that need to be tested:
1. Test customer created on master and migrated with this PR on test mode 
2. Test customer created on master and migrated with this PR on live mode 
3. Live customer created on master and migrated with this PR on live mode 
4. Live customer created on master and migrated with this PR on test mode

For each scenario we should:

*If you don't want to make a purchase, just add a payment method using the My account page. It should be enough to create the customer and check if the payment methods are migrated successfully. Don't forget to at least add something to the cart and go to the checkout page to see the shopper's payment methods.*

1. Checkout `master`
2. Login with a WC User
2. Make a purchase or add a payment method
3. Switch to this branch
4. Apply the test mode change (if applicable) and go to the checkout page
5. Assert that it doesn't break when the current mode is different than the one in which the customer was created
6. Assert that the customer's payment methods are displayed correctly. They should only be displayed in the mode they were created
    - If the migration happens in the same mode as the customer creation, the customer and their tokens should be automatically imported
    - If the migration happens in a different mode, then the customer and their tokens should be migrated the right mode. Unless a user already has a `customer_id` in the current mode, no payment methods should be displayed
7. If live/test mode were switched, make another purchase or add a payment method
8. Once a single WC User has both a test and live customer, go to the checkout page in test and in live modes
9. Assert that the cards displayed on checkout are the ones for the current mode

**To test the add cards from My account fix, don't load the checkout page before adding a new card, as that would trigger the migration**

-------------------

- [ ] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

luizreis committed:
Store test and live mode customer IDs in separate options

When a merchant toggles test mode on or off, the current customer saved
for a user is invalidated. This leads to that customer being recreated,
thus losing their saved methods and associated data on the platform.

This commit takes test mode into account when saving the customer ID
option, so a store can have both of those for the same user, preventing
errors when fetching a live customer in test mode and vice-versa.

luizreis committed:
Add customer ID to token when creating it

luizreis committed:
Extract payment method import logic to import_customer_tokens

luizreis committed:
Remove unavailable tokens before displaying them

Tokens attached to a user might belong to different customers, for
example, live and test mode ones. This commit filters the gateway
token list and displays only the tokens that belong to the current
customer, to avoid issues during checkout.

luizreis committed:
Add test mode information to user tokens

If a token is used in an off-session payment, the request should
respect the environment in which the token was created.

This commit adds the test mode information to the token, so it
can be referenced later when using it.

dwainm commented on includes/class-wc-payments-customer-service.php:
```diff
@@ -236,4 +236,11 @@ private function recreate_customer( $user, $name, $email ) {
+	 * Returns the name of the customer option meta, taking test mode into account.
+	 */
+	private function get_customer_id_option() {
+		return self::WCPAY_CUSTOMER_ID_OPTION . ( WC_Payments::get_gateway()->is_in_test_mode() ? '_test' : '' );
```
This could be simpler./ 
```suggestion
if ( WC_Payments::get_gateway()->is_in_test_mode() ) {
  return self::WCPAY_CUSTOMER_ID_OPTION . '_test' ;
} eslse {
 return  self::WCPAY_CUSTOMER_ID_OPTION;
}
```

luizreis replied:
Thanks for catching that. I've simplified it in https://github.com/Automattic/woocommerce-payments/pull/918/commits/4506fd3a0510488b01483366029ff360d04adf49.

dwainm commented on tests/test-class-wc-payments-customer-service.php:
```diff
@@ -100,6 +114,28 @@ public function test_create_customer_for_user() {
+		$customer_id = $this->customer_service->create_customer_for_user( $user, 'Test User', 'test.user@example.com' );
+
+		$this->assertEquals( 'cus_test12345', $customer_id );
+		$this->assertEquals( 'cus_test12345', get_user_option( '_wcpay_customer_id_test', $user->ID ) );
```
It would be nice to make sure the `live` id was not set for this user.

luizreis replied:
Good idea! Added the check for both tests in https://github.com/Automattic/woocommerce-payments/pull/918/commits/6cfa1142b6460b901a0f8d52a98f31b254cc6e89.

dwainm reviewed (commented):
Just had a quick scan and all looks good. Added some minor things to consider. Let me know when you need a full review.

luizreis committed:
Improve customer creation tests metadata assertions

Assert that the live mode option is not changed when creating a
test customer and vice versa.

luizreis committed:
Simplify get_customer_id_option code

luizreis committed:
Migrate old tokens optimistically when fetching them

This commit adds a customer and test_mode flag to tokens that don't have a
customer in woocommerce_get_customer_payment_tokens, taking into account
the current customer and test mode option.

luizreis committed:
Handle resource_missing exception in woocommerce_get_customer_payment_tokens

This commit removes the handling of resource_missing exceptions from
WC_Payments_Customer_Service->get_payment_methods_for_customer and moves
it to WC_Payments_Token_Service->woocommerce_get_customer_payment_tokens.

This enables us to fix the user and token data before the customer is
recreated, avoiding the loss of customer data due to losing the link
between the WC user and the Stripe customer.

woocommerce_get_customer_payment_tokens

luizreis committed:
Use token's test mode in remove_unavailable_tokens

luizreis committed:
Add change_customer_mode method to WC_Payments_Customer_Service

This method allows us to move a customer from test to live mode and vice-versa.

luizreis committed:
Migrate customers from live to test mode if resource_missing

Users created before the test and live mode decoupling commit might be
saved in the wrong meta. This commit checks if the resource_missing
error message indicates that the customer exists in the other mode
and moves it to the right meta in that case.

luizreis committed:
Use gateway_test_mode_provider in test_add_token_to_user

luizreis committed:
Check if customer_id is not null before migrating and importing tokens

luizreis committed:
Update customer meta key for live mode to include suffix '_live'

The original approach of keeping the same name makes it harder to migrate
existing customers. Now, that we're migrating customers that don't match
the current mode, we can move live mode customers without risk of losing
their data.

luizreis committed:
Migrate old customer meta to new ones based on current gateway mode

This commit modifies get_customer_id_by_user_id to update the new customer ID
options if a customer ID is found in the deprecated key, and return its value.

luizreis committed:
Do not update tokens that already match customer and test_mode when migrating

luizreis committed:
Save strval for test_mode metadata to avoid issues with tests

luizreis commented:
> Just had a quick scan and all looks good. Added some minor things to consider. Let me know when you need a full review.

Thanks, @dwainm! I think this is in good shape, but I haven't tested it extensively yet. I still need to test it in a live site and make sure the mode switches work as expected. I'll do that tomorrow first thing, but feel free to give this a try if you have some time and can get to this before I do.

luizreis committed:
Move customer_exists_in_other_mode to API client

This commit moves customer_exists_in_other_mode from the Token Service
to the API client class, as it fits better in the latter.

luizreis committed:
Make migrate_existing_tokens public

luizreis committed:
Fix customer test mode when adding card from my account

When creating a setup intent, it is possible to get the same error we
get when fetching the payment methods for a customer. If the customer
belongs to the other mode than the current one, we need to migrate it
as well. This commit adds the check and migration when adding cards.

luizreis reviewed (commented):
I've added a couple of comments to the code to explain the reasoning behind some changes. I've also added some details to the commit messages.

This PR got a little too broad and complicated to avoid losing the link between the WC users and Stripe customers. If we want to ship a simpler version to avoid having all of this "legacy" code, we could assume that all customers are live mode and discard them if they're not. I'm not sure that's worth it at this point, though.

So, I'd appreciate a few pairs of eyes here to see if the changes make sense and get some opinions if we should go this far to avoid losing test customers information.

luizreis commented:
Closing this PR after splitting into #972 and #973, and after we decided to simplify the migration, taking a live-mode optimistic approach and letting test customers get deleted once the customer is updated on checkout (paJDYF-QQ-p2).
